pkg/loki/modules.go

```diff
@@ -1816,57 +1815,29 @@ func (t *Loki) initBlockBuilder() (services.Service, error) {
 	// TODO(owen-d): perhaps refactor to not use the ingester config?
 	id := t.Cfg.Ingester.LifecyclerConfig.ID
 
-	ingestPartitionID, err := partitionring.ExtractIngesterPartitionID(id)
-	if err != nil {
-		return nil, fmt.Errorf("calculating block builder partition ID: %w", err)
-	}
-
-	reader, err := partition.NewKafkaReader(
-		t.Cfg.KafkaConfig,
-		ingestPartitionID,
-		logger,
-		prometheus.DefaultRegisterer,
-	)
-	if err != nil {
-		return nil, err
-	}
-
-	offsetManager, err := partition.NewKafkaOffsetManager(
-		t.Cfg.KafkaConfig,
-		ingestPartitionID,
-		id,
-		logger,
-		prometheus.DefaultRegisterer,
-	)
-	if err != nil {
-		return nil, err
-	}
-
-	controller, err := blockbuilder.NewPartitionJobController(
-		reader,
-		offsetManager,
-		t.Cfg.BlockBuilder.Backoff,
-		logger,
-	)
-
+	objectStore, err := blockbuilder.NewMultiStore(t.Cfg.SchemaConfig.Configs, t.Cfg.StorageConfig, t.ClientMetrics)
 	if err != nil {
 		return nil, err
 	}
 
-	objectStore, err := blockbuilder.NewMultiStore(t.Cfg.SchemaConfig.Configs, t.Cfg.StorageConfig, t.ClientMetrics)
-	if err != nil {
-		return nil, err
+	readerFactory := func(partitionID int32) (partition.Reader, error) {
+		return partition.NewKafkaReader(
```

Comment:
This will panic b/c it's creating new clients each time, each which use the same metrics namespacing internally. Instead, we could create a single client which creates cheap copies via a `WithPartition(x) -> Self` or similar.

Reply:
not sure if it is safe to make copies of `kgo.Client`. we'd need separate instances of it as we mutate it while setting offset for consumption.

working around this by registering metrics only once https://github.com/grafana/loki/pull/15224/commits/786186aa092603ce3ed50e3dbe033a3b54e3764e.